cmd/yggd/exec.go

```diff
@@ -1,191 +1,74 @@
-		log.Tracef("delaying worker start for %v...", delay)
-		time.Sleep(delay)
-	}
-
 	stdout, err := cmd.StdoutPipe()
 	if err != nil {
-		log.Errorf("cannot connect to stdout: %v", err)
-		return
+		return fmt.Errorf("cannot connect to stdout: %v", err)
 	}
 
 	stderr, err := cmd.StderrPipe()
 	if err != nil {
-		log.Errorf("cannot connect to stderr: %v", err)
-		return
+		return fmt.Errorf("cannot connect to stderr: %v", err)
 	}
 
 	if err := cmd.Start(); err != nil {
-		log.Errorf("cannot start worker: %v: %v", file, err)
-		return
-	}
-	log.Debugf("started process: %v", cmd.Process.Pid)
-
-	go func() {
-		scanner := bufio.NewScanner(stdout)
-		for scanner.Scan() {
-			log.Tracef("[%v] %v", file, scanner.Text())
-		}
-		if err := scanner.Err(); err != nil {
-			log.Errorf("cannot read from stdout: %v", err)
-		}
-	}()
-
-	go func() {
-		scanner := bufio.NewScanner(stderr)
-		for scanner.Scan() {
-			log.Errorf("[%v] %v", file, scanner.Text())
-		}
-		if err := scanner.Err(); err != nil {
-			log.Errorf("cannot read from stderr: %v", err)
-		}
-	}()
-
-	pidDirPath := filepath.Join(yggdrasil.LocalstateDir, "run", yggdrasil.LongName, "workers")
-
-	if err := os.MkdirAll(pidDirPath, 0755); err != nil {
-		log.Errorf("cannot create directory: %v", err)
-		return
+		return fmt.Errorf("cannot start process: %v: %v", file, err)
 	}
 
-	if err := ioutil.WriteFile(filepath.Join(pidDirPath, filepath.Base(file)+".pid"), []byte(fmt.Sprintf("%v", cmd.Process.Pid)), 0644); err != nil {
-		log.Errorf("cannot write to file: %v", err)
-		return
+	if started != nil {
+		go started(cmd.Process.Pid, stdout, stderr)
 	}
 
-	go watchProcess(cmd, delay, died)
-}
-
-func watchProcess(cmd *exec.Cmd, delay time.Duration, died chan int) {
-	log.Debugf("watching process: %v", cmd.Process.Pid)
-
-	state, err := cmd.Process.Wait()
-	if err != nil {
-		log.Errorf("process %v exited with error: %v", cmd.Process.Pid, err)
-	}
-
-	died <- state.Pid()
-
-	if state.SystemTime() < time.Duration(1*time.Second) {
-		delay += 5 * time.Second
-	}
-	if delay >= time.Duration(30*time.Second) {
-		delay = -1
-	}
-
-	go startProcess(cmd.Path, cmd.Env, delay, died)
-}
-
-func killProcess(pid int) error {
-	process, err := os.FindProcess(int(pid))
-	if err != nil {
-		return fmt.Errorf("cannot find process with pid: %w", err)
-	}
-	if err := process.Kill(); err != nil {
-		log.Errorf("cannot kill process: %v", err)
-	} else {
-		log.Infof("killed process %v", process.Pid)
-	}
 	return nil
 }
 
-func killWorker(pidFile string) error {
-	data, err := ioutil.ReadFile(pidFile)
+// waitProcess finds a process with the given pid and waits for it to exit.
+// If died is not nil, it is invoked on a goroutine when the process exits.
+func waitProcess(pid int, died func(pid int, state *os.ProcessState)) error {
+	process, err := os.FindProcess(pid)
 	if err != nil {
-		return fmt.Errorf("cannot read contents of file: %w", err)
+		return fmt.Errorf("cannot find process with pid: %v", err)
```

Comment:
In this function(startWorker) all the logs are well defined, but in some case like this we log the PID, in other the program, maybe will be good to add some metadata like log.WithFileds(programm, programme), so you know what worker failed, emit log, etc..

Reply:
It looks like the logging could be improved overall, so I'll take a pass over the whole branch.

Reply:
I'm passing over the logging and error messages, and found a few improvements. Like here, it should include the PID as well as the underlying error returned from FindProcess. However at this layer, waitProcess knows nothing else about the program it is told to wait for. It's given a PID; it has no awareness of a worker or a worker config.